IterDomain resize for pad, cat, slice

naoyam opened the pull request:
This PR adds codegen support of `pad`, `slice` and `cat`. All of them are represented using a new IterDomain expression, `Resize`. Please see test_gpu_resize.cpp for concrete examples. 

The pointwise, reduction and normalization schedulers should be able to work with resized domains. Transformation and parallelization propagations are done while skipping, or forwarding, resized domains. So, for example:

```
tv0: [I0, I1]
tv1: [I0+2, I1] // axis 0 is padded by one at each of its left and right sides

tv1->split(0, 4);
tv1: [ceilDiv(I0 + 2, 4), 4, I1]

// After propagating tv1 to tv0
tv0: [ceilDiv(I0, 4), 4, I1]
```

And for parallelization:
```
tv1->axis(0)->parallelize(ParallelType::BIDx);
tv1: [bidx(ceilDiv(I0 + 2, 4)), 4, I1]

// After propagating tv1 to tv0
tv0: [bidx(ceilDiv(I0, 4)), 4, I1]
```

Unlike these propagations, inlining is not allowed between these domains with different extents, at least at this moment. So, `inlineMost` would not change the CA position of `tv0` from 0.

The difference of honoring (i.e., inlining) or ignoring resize (transform and parallel propagations) requires different behavior in some of the underlying replay primitives such as `BestEffortReplay`. As you can see, a new option, `skip_resize`, is added, which behaves similar to `skip_swizzle`. This unfortunately further makes them error prone due to optional boolean flags ([here's a recent bug we found](#2531)). I'm already prototyping a cleanup PR to refactor these options so that they are more explicitly configured without little ambiguity.

I'm also thinking about adding two more IterDomain expressions, `gather` and `scatter`. With `resize` and `gather`, we should be able to implement the PyTorch gather.

Another thing I'm really interested in trying out is to use `slice` and `cat` to allow multiple memory spaces to be used for a single tensor in persistent kernels. I believe this should be possible, but it would need some more work.

TODOs
- [x] Add validation of resize usage. Must be within root and rfactor domains when given to GpuLower
- [ ] Fix string matching errors of generated code with some of the C++ tests

TODOs likely not in this PR
- [ ] Support non-zero constant padding

Commit:
WIP

Commit:
WIP: expand

Commit:
WIP

Commit:
WIP

Commit:
v1

Commit:
wip

Commit:
tests

Commit:
WIP: PadOp

Commit:
checkpoint: decided to try rfactor

Commit:
pad test

Commit:
computeAt

Commit:
parallelize

Commit:
concat

Commit:
cat tests

Commit:
WIP: slice

Commit:
bug fix

Commit:
cleanup

Commit:
Merge branch 'devel' into catop

Commit:
Merge branch 'devel' into catop

Commit:
Renamed expand to resize

Commit:
Merge branch 'devel' into catop

Commit:
cleanup

Commit:
Merge branch 'devel' into catop

Commit:
cleanup

Commit:
cleanup

Commit:
remove TensorView::expand

Commit:
cleanup

Commit:
fix

Commit:
cleanup

naoyam commented on third_party/nvfuser/csrc/lower_shift.cpp:
```diff
@@ -438,25 +438,20 @@ void HaloInfo::build(TensorDomain* td) {
-      }
     } else {
-      TORCH_INTERNAL_ASSERT(false, "Unsupported expr: ", expr);
+      // Assume no halo
```
Nothing changed. Just make the logic for non-supported exprs more generic

naoyam commented on third_party/nvfuser/csrc/lower_shift.cpp:
```diff
@@ -832,28 +827,23 @@ std::unordered_map<IterDomain*, Val*> HaloInfo::buildConcreteHaloExtentMap(
-          0);
     } else {
-      TORCH_INTERNAL_ASSERT(false, "Unsupported expr: ", expr);
+      // Halo not yet supported, just set the width to zero at the moment.
```
Same here

Commit:
cleanup

Commit:
cleanup

Commit:
cleanup

Commit:
update

Commit:
cleanup

Commit:
fix

Commit:
Merge branch 'devel' into iter_domain_resize

Commit:
test cleanup

Commit:
cleanup

Commit:
unswitch fix

Commit:
updates for resize

Commit:
more tests

Commit:
add scheduler tests for cat

Commit:
Don't propagate resize if not safe

Commit:
update

Commit:
WIP: scheduler support

Commit:
cleanup

Commit:
fix

Commit:
fix

Commit:
Softmax + slice working

Will stop propagating resize

Commit:
cleanup, should be more robust

Commit:
cleanup

Commit:
All tests pass

Commit:
fix

Commit:
cleanup

Commit:
cleanup

Commit:
fix

Commit:
cleanup

Commit:
Merge branch 'devel' into iter_domain_resize

naoyam commented on third_party/nvfuser/csrc/index_compute.cpp:
```diff
@@ -1336,97 +1367,10 @@ std::vector<Val*> Index::getGlobalProducerStridedIndices(
     const std::unordered_map<IterDomain*, Val*>& override_index) {
   FUSER_PERF_SCOPE("GpuLower::Lower::getGlobalProducerIndex");
 
-  // Replay producer to look like consumer so we can index on producer since
```
Just refactored this part as `getProducerRootIndices` as it is done for consumer indexing with `getConsumerRootIndices`.

naoyam commented on third_party/nvfuser/csrc/index_compute.cpp:
```diff
@@ -1904,6 +1827,152 @@ std::vector<Val*> Index::getRootIndices(
   return root_inds;
 }
 
+std::vector<Val*> Index::getProducerRootIndices(
```
Just extracted out of `getGlobalProducerStridedIndices`

naoyam commented on third_party/nvfuser/csrc/index_compute.cpp:
```diff
@@ -2862,18 +2990,13 @@ std::vector<RootPredicateInfo> Index::getReferenceRootPredicates(
 
     // Build predicates for start positions as:
     //   start_index + start_offset >= 0
-    auto start_offset = simplifyStartOffset(info.start_offset_);
```
`simplifyStartOffset` needs to be updated to support predicates with resize. I just removed it as the expr simplifier can do almost the same job. It doesn't remove some halo-specific predicates, but that's not really important right now, and we should extend the expr simplifier anyway instead of extending `simplifyStartOffset`.

zasdfgbnm replied:
Will this PR help with the halo-specific predicates? https://github.com/csarofeen/pytorch/pull/2541

I am just asking. I am OK with removing `simplifyStartOffset` even if #2541 does not help.

naoyam replied:
Yes, I think #2541 would help. Specifically, we are using `max` and `min`, but in some cases we can assume which one should be the maximum or the minimum, so that's I think we could do with `prove::lessThan` etc. Thanks for extending the expr simplifier. That's really helpful as it allows us to focus on the correctness.

naoyam commented on third_party/nvfuser/csrc/scheduler/pointwise_utils.cpp:
```diff
@@ -5,7 +5,7 @@ namespace nvfuser {
 
 DomainMap::DomainMap(Fusion* fusion) : fusion_(fusion), ca_map_(fusion) {
-  view_tvs_ = scheduler_utils::getViewTVs(fusion);
+  tvs_with_rfactor_ = scheduler_utils::getTVsWithRFactor(fusion);
```
Changed to include rfactor domains with `resize`.

naoyam commented on third_party/nvfuser/csrc/scheduler/utils.cpp:
```diff
@@ -246,7 +247,8 @@ void parallelizeAllLike(
        it++) {
-    auto ca_id = ca_map.getConcreteMappedID(*it, IdMappingMode::PERMISSIVE);
+    auto ca_id =
+        ca_map.getConcreteMappedID(*it, IdMappingMode::PERMISSIVE_RESIZE);
```
Changed it to allow propagation across resized domains. May not be ideal, but I think it's a reasonable default behavior.

Commit:
cleanup

Commit:
cleanup

Commit:
cleanup

Commit:
Merge branch 'devel' into iter_domain_resize

Commit:
cleanup

Commit:
cleanup

Commit:
Merge branch 'devel' into iter_domain_resize

Commit:
cleanup

naoyam commented on third_party/nvfuser/test/test_gpu2.cpp:
```diff
@@ -4340,8 +4340,10 @@ TEST_F(NVFuserTest, FusionVectorizeMisalignedStride_CUDA) {
-  at::Tensor t0 = at::randn({bx, by}, options).index({"...", Slice(3)});
-  at::Tensor t1 = at::randn({bx, by}, options).index({"...", Slice(3)});
+  at::Tensor t0 =
+      at::randn({bx, by}, options).index({"...", at::indexing::Slice(3)});
```
Since I added `Slice` in our namespace, I needed to change it to use the full namespace here.

zasdfgbnm commented on third_party/nvfuser/csrc/index_compute.h:
```diff
@@ -321,11 +322,19 @@ class Index {
   static std::vector<Val*> getStrides(const TensorView* tv);
 
   // get the root indices of a tensor used for the index lowering
```
```suggestion
  // get the root indices of a consumer tensor
```

zasdfgbnm commented on third_party/nvfuser/csrc/ir_internal_nodes.h:
```diff
@@ -1465,6 +1465,14 @@ class TORCH_CUDA_CU_API IterDomain : public Val {
+      IterDomain* in,
+      Val* left_expansion,
+      Val* right_expansion,
+      bool mark_as_rfactor = false);
```
Do we need to mark it as rfactor domain for some case and not do so for other cases? It would be good to add a comment about when `mark_as_rfactor` should be true and when `mark_as_rfactor` should be false.

naoyam replied:
Added a comment.

zasdfgbnm commented on third_party/nvfuser/csrc/ir_internal_nodes.h:
```diff
@@ -2125,6 +2133,46 @@ class TORCH_CUDA_CU_API Swizzle2D : public Expr {
+  }
+
+  Val* leftExpand() const {
+    return attributeVal(0);
```
Should we make left expand and right expand inputs instead of attributes? If they are not inputs of the expr,  then when they are fusion inputs and the fusion is segmented, the segments will be missing fusion inputs. I had a similar issue when I was working on tensor factories.

naoyam replied:
Since this is an IterDomain expression, I don't think it affects the fusion segmentation, where only TensorView and scalar expressions are considered.

The `factor` parameter of the `Split` expression is also an attribute.

That said, I'm not sure which would make more sense if it's an attribute or an input. I think I just followed what `Split` does. Since the output IterDomain does depend on the expansion parameters, it seems they should be considered inputs.

zasdfgbnm replied:
Yeah, as long as the test mentioned in https://github.com/csarofeen/pytorch/pull/2480#discussion_r1122682789 works fine, there should be no problem of having it as an attribute.

naoyam replied:
For `Resize`, marking the expand parameters as inputs results in other failures. For example, 

https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/csrc/ir_nodes.cpp#L2273-L2286

here, we assume that all inputs of an IterDomain can be found in the root domain vector, which isn't the case if we set the expand parameters also as inputs. I think some parts of the code, especially older ones, have implicit assumptions that inputs to IterDomain expressions are also IterDomain. 

I'll leave it as is for now.

zasdfgbnm replied:
Thanks for checking. It makes sense to leave it as is.

zasdfgbnm commented on third_party/nvfuser/test/test_gpu_resize.cpp:
```diff
@@ -0,0 +1,1546 @@
+  TORCH_CHECK(ref.equal(cg_outputs[0]));
+}
+
+TEST_F(NVFuserTest, FusionResizePadReduceScheduler1_CUDA) {
```
Could you add another test case that satisfies the following:
1. There are multiple pads
2. Different pads are segmented into different fusions
3. `left_pad` and `right_pad` are fusion inputs

This will check if the segmenter can handle your expr definition correctly, and if you have to make these scalars expr inputs or if it is OK to be attributes.

naoyam replied:
Pushed a new commit: 2a2e18942361de1fad0c95ec5277c4173119ad1e

Added a test as you suggested. Turned out I missed to mark the width inputs of `PadOp` as inputs. That's fixed now, and the test works fine. Please let me know if this test looks sufficient.

zasdfgbnm replied:
That test looks sufficient.

zasdfgbnm commented on third_party/nvfuser/csrc/ir_nodes.cpp:
```diff
@@ -2996,4 +3074,209 @@ c10::optional<ParallelType> NamedScalar::getParallelIndex() const {
+  addOutput(out);
+  addInput(inp);
+  for (auto width : pad_widths) {
+    addAttribute(width);
```
Same here, should we make these widths inputs instead of attributes? If they are not inputs of the expr, then when they are fusion inputs and the fusion is segmented, the segments will be missing fusion inputs. I had a similar issue when I was working on tensor factories.

naoyam replied:
Thanks for the comment. Fixed.

zasdfgbnm commented on third_party/nvfuser/csrc/compute_at_map.h:
```diff
@@ -46,6 +46,10 @@ namespace nvfuser {
 //   Map all iteration domains
 //   Always contain root mappings (otherwise they could have been forwarded in
 //   broadcast)
+// IdMappingMode::PERMISSIVE_RESIZE
```
Not for this PR, but it looks like whenever we need a new op, we will add a new type of permissive map. I don't think this design scales well. For example, if we have `PERMISSIVE_X`, `PERMISSIVE_Y`, `PERMISSIVE_Z`, ... Will there be cases where we want a permissive map that needs map `X` and `Z` but not `Y`?

naoyam replied:
Yeah, agreed. I think #2316 by @csarofeen would make this more manageable. For the meantime, I believe it's also good to exercise the indexing map. That would allow us to understand what capability the indexing map would need to have.

zasdfgbnm commented on third_party/nvfuser/csrc/compute_at_map.cpp:
```diff
@@ -433,7 +444,14 @@ void IterDomainGraph::build(Fusion* fusion) {
         const auto permissive_disjoint_sets =
-            BestEffortReplay::replayPasC(p_tv, c_tv, -1, pairwise_map)
+            BestEffortReplay::replayPasC(
+                p_tv, c_tv, -1, pairwise_map, true, true, false)
```
nit: Comment what these `true` and `false` are.

naoyam replied:
Added comments. I'll also clean up these boolean flags.

zasdfgbnm commented on third_party/nvfuser/csrc/compute_at_map.cpp:
```diff
@@ -433,7 +444,14 @@ void IterDomainGraph::build(Fusion* fusion) {
+        // Permissive-Resize map allows mappings of resize inputs and outputs
+        const auto permissive_resize_disjoint_sets =
+            BestEffortReplay::replayPasC(
+                p_tv, c_tv, -1, pairwise_map, true, true, true)
```
nit: Comment what these `true` and `false` are.

zasdfgbnm commented on third_party/nvfuser/csrc/compute_at_map.cpp:
```diff
@@ -503,6 +521,17 @@ void IterDomainGraph::build(Fusion* fusion) {
+        // nothing to do for the loop map
+        for (auto& dset : permissive_resize_disjoint_sets.disjointSets()) {
+          auto& vec = dset->vector();
+          for (auto i : c10::irange(vec.size())) {
```
Do we need to update `producers_` and `consumers_`?

naoyam replied:
🤔

naoyam replied:
They should be based on the most permissive mappings, so yes, they should be created using the permissive-resize map. See 27c867f11f81b096066d5f526be0526b2ea9640d.

This would also be an issue if we would end up having multiple maps that there may not be ordering of permissiveness.

zasdfgbnm commented on third_party/nvfuser/csrc/ir_internal_nodes.h:
```diff
@@ -2181,4 +2229,121 @@ class TORCH_CUDA_CU_API NamedScalar : public Val {
+  Val* step = nullptr;
+};
+
+class TORCH_CUDA_CU_API SliceOp : public Expr {
```
Should slice also be an IterDomain expr?

naoyam replied:
I don't think we need a different IterDomain expr as it can be expressed with `Resize`. 

However, there could be some situations where we may want to differentiate or keep information about how an IterDomain is resized. For example, when `Resize` is used for slicing, it is guaranteed that the output IterDomain has a smaller extent than the input, whereas it's the opposite when used for padding. Such information may be useful in some cases.

zasdfgbnm commented on third_party/nvfuser/csrc/ir_internal_nodes.h:
```diff
@@ -2181,4 +2229,121 @@ class TORCH_CUDA_CU_API NamedScalar : public Val {
+    return attribute(0)->as<Attribute<int>>()->value;
+  }
+
+  //! The index val that determines which input tensor should be used
+  //! to fill the particular output position of this expression. Only
+  //! valid after indexing
+  Val* getConcatenatedDomainIndex() const;
+
+  //! Gets a Bool indicating if the input tensor specified by
+  //! tensor_idx should be used to fill the output tensor. Only valid
+  //! with the Kernel container
+  Bool* getPred(int input_idx) const;
```
Instead of putting lowering information here and handle it in `codegen.cpp`, does it make sense to lower cat to a sequence of ternary where op, or even crazier, `kir::IfThenElse`?

zasdfgbnm replied:
Sorry, didn't see your comment in https://github.com/csarofeen/pytorch/pull/2480/files#diff-66e2e03e8b5e393d98fbcde0edfc1ef72ba81526bfaa8ef5a316907ebcff2892R1440-R1443

naoyam replied:
Yeah, I think this is reasonable. We could extend `kir::IfThenElse` to have `else if` clauses as well, but that seems overkill if the only use case is here.

zasdfgbnm commented on third_party/nvfuser/csrc/ir_nodes.cpp:
```diff
@@ -2127,6 +2128,52 @@ std::pair<IterDomain*, IterDomain*> IterDomain::swizzle(
+    Val* right_expansion,
+    bool mark_as_rfactor) {
+  TORCH_CHECK(
+      left_expansion->isScalar(),
```
Should we also check if `left_expansion` is int?

zasdfgbnm commented on third_party/nvfuser/csrc/ir_nodes.cpp:
```diff
@@ -2915,6 +2962,37 @@ std::string Swizzle2D::toInlineString(int indent_size) const {
+  std::stringstream ss;
+  ss << "Resize: ";
+  ss << in()->toString();
+  ss << " by " << leftExpand()->toString() << " and "
+     << rightExpand()->toString();
```
Should we use `toInlineString` here? For example, if `leftExpand()` is `ceilDiv(5, 3)`,`toString` will just print something like `i123`?

zasdfgbnm commented on third_party/nvfuser/csrc/ir_nodes.cpp:
```diff
@@ -2996,4 +3074,209 @@ c10::optional<ParallelType> NamedScalar::getParallelIndex() const {
+  TORCH_INTERNAL_ASSERT(pad_idx >= 0 && pad_idx < num_dims);
+
+  return std::make_pair(
+      attribute(pad_idx)->as<Val>(), attribute(pad_idx + 1)->as<Val>());
```
Why do I think this should be `2 * pad_idx` and `2 * pad_idx + 1`. Do I understand the semantics wrong?

zasdfgbnm commented on third_party/nvfuser/csrc/ir_nodes.cpp:
```diff
@@ -2996,4 +3074,209 @@ c10::optional<ParallelType> NamedScalar::getParallelIndex() const {
+  std::vector<Slice> ranges(ndims);
+  for (const auto i : c10::irange(ndims)) {
+    ranges.at(i) = Slice{
+        .start = input(input_offset + i * 3)->as<Val>(),
+        .stop = input(input_offset + i * 3 + 1)->as<Val>(),
+        .step = input(input_offset + i * 3 + 2)->as<Val>()};
```
I think inputs are already `Val`?
```suggestion
        .start = input(input_offset + i * 3),
        .stop = input(input_offset + i * 3 + 1),
        .step = input(input_offset + i * 3 + 2)};
```

naoyam replied:
Ah, right. I guess this is because I originally set them as attributes. Will clean it up.

zasdfgbnm commented on third_party/nvfuser/csrc/ir_internal_nodes.h:
```diff
@@ -2125,6 +2133,46 @@ class TORCH_CUDA_CU_API Swizzle2D : public Expr {
 };
 
+//! IterDomain expression to resize
+class TORCH_CUDA_CU_API Resize : public Expr {
```
Just some random thoughts: Can `Resize` only be used for tensor definition? Can it be used for scheduling as well? For example, if I have horizontal fusion `T1 = T0, T3 = T2`, then I can resize the smaller one to the larger one?

naoyam replied:
It can be used, but since it changes the range where the tensor is defined, it can result in an invalid program. And in fact, it is used to transform a producer tensor like a consumer when indexing the producer.

For example,

```
// tv0: [I0]
auto tv1 = pad(tv0, {1, 1});
// tv1: [I1]
```

In this case, `tv1` has a rfactor domain with `resize`:
```
I1 = resize(I0, {1, 1});
```

when indexing `tv0` as the producer of `tv1`, this `resize` expression is replayed on `tv0`, making `tv0` has the same extent as `tv1`. Since this transformation of `tv0` doesn't affect how `tv0` is defined, it doesn't affect the correctness of the program.

csarofeen replied:
It seems more like a broadcast in this sense than something like a split/merge. i.e. it's really a compute definition change, not just a schedule change.

It would also be hard to know what value to initialize the new entries to in the larger dimension.

naoyam replied:
>  it's really a compute definition change, not just a schedule change.

Yes, that's why for consumer tensors, resize should only be used as rfactor. A producer tensor can be resized as demanded by its consumers. We resize a producer so that it has the same extent as its consumer. Once a producer domain has the same extent, we can propagate the consumer index to the producer.

zasdfgbnm replied:
Isn't non-divisible split also like compute definition change? `split(I{2}, 5) -> I{1}, I{5}`, they both effectively change the extent of the original ID, and I don't expect non-divisible split to be super different from resize.

naoyam replied:
That's a good point. Yes, it does change the extent. It won't matter for allocation correctness since we just overallocate. For indexing, what we are doing specifically for non-divisible splits basically means indexing is always done with IterDomains before non-divisible splits.

zasdfgbnm commented on third_party/nvfuser/csrc/scheduler/utils.cpp:
```diff
@@ -2116,6 +2141,9 @@ DisjointSets<IterDomain*> disjointViewSets(Fusion* fusion) {
         auto split = expr->as<Split>();
         disjoint_view_ids.mapEntries(split->in(), split->inner());
```
Consider renaming this variable to `disjoint_rfactor_ids`, and this function to `disjointRFactorSets `

naoyam replied:
Done: c4d6b39c0a5559696e6c385196a44eca9b7a4b2a

zasdfgbnm commented on third_party/nvfuser/csrc/transform_iter.cpp:
```diff
@@ -172,6 +172,38 @@ void ReplayTransformations::handle(Swizzle2D* swizzle_2d) {
+    }
+  }
+
+  auto mapped = (*it).second;
```
nit:
```suggestion
  auto mapped = it->second;
```

naoyam replied:
Done. Also changed the same patterns in this file.

zasdfgbnm reviewed (commented):
Not done yet, posting some of my comments.

zasdfgbnm commented on third_party/nvfuser/csrc/index_compute.cpp:
```diff
@@ -563,8 +563,38 @@ void IndexCompute::handle(Swizzle2D* swizzle_2d) {
+  auto zero = GpuLower::current()->kernel()->zeroVal();
+
+  if (isZero(out_id)) {
+    index_map_[in_id] = zero;
```
I don't understand. Why this is not `neg(resize->leftExpand())`?

naoyam replied:
Actually, when this happens, indexing the input ID is invalid. Being zero effectively means that all of the producer and consumer tensors access the domain with the same index, and there should be no such case with resize, at least as this point. For a producer tensor, since inlining is not allowed with resized domains, out_id should never be a zero domain. For a consumer tensor, resize should only appear as part of the rfactor transformations, and since indexing is done with the rfactor root domain, resize should never be involved.

Updated: 622ee939a5038f4d86dcf496cf371e313073051c

zasdfgbnm commented on third_party/nvfuser/csrc/lower_unroll.cpp:
```diff
@@ -298,6 +311,30 @@ bool UnrollPass::canOmitElseClause(kir::ForLoop* fl) {
+        resize_exprs.end(),
+        std::back_inserter(resize_expr_inputs),
+        [](Expr* resize_expr) { return resize_expr->input(0); });
+    auto dep_exprs = DependencyCheck::getAllExprsBetween(
```
I don't think this will cover all dependencies. Some lowering passes create shallow copy, so the exprs in the loop nest are not guaranteed to have the same pointer as those in the fusion. I think you should check inputs/outputs instead of expr pointers.

naoyam replied:
I'm not sure what you mean by:

> so the exprs in the loop nest are not guaranteed to have the same pointer as those in the fusion.

Are you saying a tensor could have multiple definitions and this dependency analysis only traverses one of the definitions, and thus missing the dependencies through the other definitions?

That's a good point, and I missed that.

naoyam replied:
I believe this fixes the issue:  b0ffb91317

zasdfgbnm replied:
Yes, your understanding is correct.

zasdfgbnm commented on third_party/nvfuser/csrc/scheduler/utils.h:
```diff
@@ -557,5 +560,19 @@ inline void rotateLoop(
 
+//! Certain tensors may need to be placed on shared or global memory
+//! due to data dependencies caused by resize operations. Create
+//! cachees of those tensors so that original operations producing
```
```suggestion
//! caches of those tensors so that original operations producing
```

zasdfgbnm commented on third_party/nvfuser/csrc/scheduler/utils.cpp:
```diff
@@ -2317,6 +2355,124 @@ bool isFastestDimReduction(TensorView* tv) {
+    }
+  }
+
+  // Iterate over required_tensors so that promotion is done in a
```
```suggestion
  // Iterate over resized_tensors so that promotion is done in a
```

zasdfgbnm commented on third_party/nvfuser/csrc/scheduler/utils.cpp:
```diff
@@ -2317,6 +2355,124 @@ bool isFastestDimReduction(TensorView* tv) {
+  }
+}
+
+void promoteProducerMemoryTypesOfResizedTensors(Fusion* fusion) {
```
Interesting. Can this function be used to do reduction-transpose and persistent-transpose fusion?

naoyam replied:
A similar analysis could work too.

zasdfgbnm commented on third_party/nvfuser/csrc/ops/alias.cpp:
```diff
@@ -371,4 +372,281 @@ TensorView* transpose(TensorView* x) {
+TensorView* pad(TensorView* inp, const std::vector<Val*>& pad_widths) {
+  const auto& inp_dom = inp->domain()->noReductions();
+
+  const auto ndims = inp->domain()->noReductions().size();
```
```suggestion
  const auto ndims = inp_dom.size();
```

zasdfgbnm commented on third_party/nvfuser/csrc/ops/alias.cpp:
```diff
@@ -371,4 +372,281 @@ TensorView* transpose(TensorView* x) {
+// output. All of the inputs to CatOp have the same shape as the
+// output shape.
+TensorView* cat(const std::vector<TensorView*>& inputs, int cat_dim) {
+  TORCH_CHECK(!inputs.empty(), "No input tensor given");
```
Consider special handling the case where `inputs.size() == 1`

zasdfgbnm commented on third_party/nvfuser/csrc/ops/alias.cpp:
```diff
@@ -371,4 +372,281 @@ TensorView* transpose(TensorView* x) {
+            TensorDomain::getContiguousContiguity(rfactor_ids)),
+        dtype);
+
+    IrBuilder::create<PadOp>(resized_inp, inputs.at(input_idx), pad_widths);
```
Why not use the `pad` function above?

naoyam replied:
Thanks for the suggestion. Done: 8ad222c0fc76c48c885bfdfba14da5d96e00ab34

zasdfgbnm commented on third_party/nvfuser/csrc/ops/alias.cpp:
```diff
@@ -371,4 +372,281 @@ TensorView* transpose(TensorView* x) {
+
+  TORCH_CHECK(ndims == static_cast<int>(ranges.size()));
+
+  auto normalize_slice_range = [](const Slice& range, Val* extent) -> Slice {
+    auto normalized_range = range;
+    if (range.start == nullptr) {
+      normalized_range.start = FusionGuard::getCurFusion()->zeroVal();
+    }
+    if (range.stop == nullptr) {
+      normalized_range.stop = extent;
+    }
+    if (range.step == nullptr) {
+      normalized_range.step = FusionGuard::getCurFusion()->oneVal();
+    }
+    return normalized_range;
+  };
```
nit:
```suggestion
  auto normalize_slice_range = [](Slice range, Val* extent) -> Slice {
    if (range.start == nullptr) {
      range.start = FusionGuard::getCurFusion()->zeroVal();
    }
    if (range.stop == nullptr) {
      range.stop = extent;
    }
    if (range.step == nullptr) {
      range.step = FusionGuard::getCurFusion()->oneVal();
    }
    return range;
  };
```

zasdfgbnm commented on third_party/nvfuser/csrc/ir_internal_nodes.h:
```diff
@@ -1465,6 +1465,14 @@ class TORCH_CUDA_CU_API IterDomain : public Val {
       bool inner_split,
       bool trim_out_of_bounds);
 
+  //! Resize an IterDomain by expanding both the left and right sides
+  //! by given widths. Optionally mark it as an rfactor domain
```
Maybe mention that `left_expansion` and `right_expansion` may be negative, and in this case, will shrink instead of expand.

zasdfgbnm commented on third_party/nvfuser/csrc/ops/alias.h:
```diff
@@ -66,4 +66,23 @@ TORCH_CUDA_CU_API TensorView* transpose(
+//! Pad a tensor by given widths of zero. Similar to torch.pad, the
+//! pad_widths vector specifies the padding widths of the innermost N
+//! dimensions, where N is half the size of the width vector. Padding
+//! is always done just by zero. TODO: Support other padding types
```
Not asking to add support for non-zero padding value in this PR, but supporting non-zero padding value is not hard:
```python
def pad(x, pad_widths, pad_value):
    mask = torch.tensor(True).expand_as(x)
    x_pad_zero = torch.pad(x, pad_widths)
    mask_pad_zero = torch.pad(mask, pad_widths)
    return torch.where(mask_pad_zero, x_pad_zero, pad_value)
```

naoyam replied:
Yeah, the constant padding should be a trivial extension. The other modes of padding, 'reflect', 'replicate' or 'circular', may not be.

csarofeen replied:
I think const case makes sense to add for now. Agree the others could be more challenging/interesting to support. I don't think they're required at the moment.

csarofeen replied:
Maybe a better way to extend this would be to make Pad its own IterDomain expr. Again, I think we can leave this to future work.

naoyam replied:
Yes, I'll add support of non-zero padding, but I prefer it as a separate PR as this is already rather large.

zasdfgbnm commented on third_party/nvfuser/csrc/transform_replay.cpp:
```diff
@@ -802,7 +826,8 @@ int TransformReplay::getMatchedLeafPosWithoutReplayPasC(
     const TensorView* producer,
-    int producer_pos) {
+    int producer_pos,
+    bool skip_resize) {
```
To be more consistent, should we add a `skip_swizzle` and do the same thing in transform propagator?

naoyam replied:
Probably yes, but I'm actually not quite sure about skipping swizzles. Looks like they are always skipped, so it's probably not necessary, but for consistency, yes. For resize, the option is necessary to have.

zasdfgbnm commented on third_party/nvfuser/csrc/transform_replay.h:
```diff
@@ -124,21 +124,30 @@ class RootDomainMap;
       int consumer_compute_at_axis,
-      bool replay_swizzle = false);
+      bool replay_swizzle = false,
+      bool replay_resize = false);
```
This is not your fault, but in `TransformReplay`, we have `replay_resize`, but in `BestEffortReplay`, we have `skip_resize`. We should consider making them use the same name.

naoyam replied:
Yes, I'll refactor them as a followup PR.

zasdfgbnm commented on third_party/nvfuser/csrc/transform_replay.h:
```diff
@@ -124,21 +124,30 @@ class RootDomainMap;
  public:
-  // Replay producer as consumer, returns {producer, producer_compute_at_axis}.
+  // Replay producer as consumer, returns {producer,
+  // producer_compute_at_axis}.
```
Why this is put into a new line?

naoyam replied:
No idea. Revered.

zasdfgbnm commented on third_party/nvfuser/csrc/transform_rfactor.cpp:
```diff
@@ -210,6 +210,33 @@ class ReplayRFactor : public ReplayTransformations {
     }
   }
 
+  void handle(Resize* resize) override {
```
Is this used? I think we never rfactor a pad/slice consumer?

naoyam replied:
You're right. I guess it was needed in some early prototype. Changed to error out if this hits. 1b5c23ccee

zasdfgbnm commented on third_party/nvfuser/csrc/contiguity.h:
```diff
@@ -55,6 +55,8 @@ class OrderedIdInformation : public OptInDispatch {
 
   void handle(Swizzle2D* swizzle) override;
 
+  void handle(Resize* expand) override;
```
Should we use names more consistently? Currently, we are using resize/expand. Should we replace all "expand" with resize?

naoyam replied:
I started with `IterDomain::expand` initially as I was only thinking about padding, but realized it's not a good name when adding support of slicing, so renamed it to `resize`. Looks like there were still some using the old name. Renamed this. I don't think there's no such name anymore.

zasdfgbnm commented on third_party/nvfuser/csrc/contiguity.cpp:
```diff
@@ -324,6 +324,48 @@ void OrderedIdInformation::handle(Swizzle2D* swizzle) {
   }
 }
 
+void OrderedIdInformation::handle(Resize* resize) {
```
Do we need this change? Will setting resize's contiguity to `false` in `TensorDomain::getContiguousContiguity` solve the problem?

csarofeen replied:
If we had a domain `[I0, resize, I1]`, and contiguity `[t, f, t]` would it be okay to merge the indexing of `I0` into the `resize`?

csarofeen replied:
This check is only for ordering of iter domains, it seems fine to extend it's analysis here consistently with Swizzle, even though swizzle can't be indexed contiguously.

zasdfgbnm replied:
I think the contiguity of resize should be similar to the contiguity of expand prior to https://github.com/csarofeen/pytorch/pull/2517, the dimension of resize and the dimension before resize should never be contiguous. And yeah, I am OK with the change here, just asking some general question.

naoyam replied:
I think what I did here is to just follow the same logic as is done for the other expressions. It's not contiguously indexed, so this isn't technically necessary.

zasdfgbnm reviewed (commented):
posting more

zasdfgbnm commented on third_party/nvfuser/csrc/index_compute.cpp:
```diff
@@ -2351,6 +2423,36 @@ std::vector<PredicateDomainInfo> getNonDivisibleConsumerDomainsToPredicate(
+       consumer_tv->domain()->domain().end()});
+
+  for (auto resize : ir_utils::filterByType<Resize>(exprs)) {
+    // If the input is a root domain, it is already predicated by the
+    // normal predicate logic
```
Hmm, in `getPredicateContigIds`, we are doing:
```C++
  // When there's a resize expr between the root and the rfactor
  // domains, predicate the rfactor domain. Otherwise, predicate the
  // root domain. The actual size of an IterDomain after resize
  // changes, and the output IterDomain needs to be used to generate
  // its predicate.
  const auto& consumer_root_domain = ir_utils::hasResizedRfactor(consumer_tv)
      ? consumer_tv->getMaybeRFactorDomain()
      : consumer_tv->getRootDomain();
```
So the normal predicate logic will not predicate `resize->in()`? Is my understanding correct?

naoyam replied:
Thanks for the comment. This is actually not necessary, and in the current implementation it has no effect as you pointed out. Just removed.

zasdfgbnm commented on third_party/nvfuser/csrc/lower_index.cpp:
```diff
@@ -1381,4 +1381,95 @@ void IndexLowering::allocateUniqueFusedReduction(
 }
 
+void IndexLowering::handle(const PadOp* pad) {
+  // Convert to a where op as:
+  // consumer[consumer_idx] = (produer_idx >= 0 && produer_idx <
+  //                           producer_extent) ?
+  //     producer[producer_idx] :
+  //     0;
```
Don't know how our predicate is generated. Is there any case that will lead to code like
```C++
if (produer_idx >= 0 && produer_idx < producer_extent) {
  consumer[consumer_idx] = (produer_idx >= 0 && produer_idx <
                            producer_extent) ?
      producer[producer_idx] :
      0;
}
```
Can this be a problem?

naoyam replied:
Predicates are done with consumers, so we should see something like:

```
if (consumer_idx >= 0 && consumer_idx < consumer_extent) {
  consumer[consumer_idx] = (produer_idx >= 0 && produer_idx <
                            producer_extent) ?
      producer[producer_idx] :
      0;
}
```

zasdfgbnm replied:
OK, then that should be fine

zasdfgbnm commented on third_party/nvfuser/csrc/lower_index.cpp:
```diff
@@ -1381,4 +1381,95 @@ void IndexLowering::allocateUniqueFusedReduction(
+}
+
+void IndexLowering::handle(const CatOp* cat) {
+  // It's possible to lower CatOp to a series of IfThenElse or Where,
+  // but that would going to look really ugly. For now, rely on
+  // CudaKernelGenerator to produce code based on the predicates
+  // genereated here.
```
Not for this PR, but I believe we should make our `codegen.cpp` smart to handle:
```C++
kir::IfThenElse pred1:
Then:
  ...
Else:
  kir::IfThenElse pred2:
  ...
  End ITE
End ITE
```
to generate code like
```C++
if (pred1) {
} else if (pred2) {
}
```
rather than
```C++
if (pred1) {
} else {
  if (pred2) {
  }
}
```

zasdfgbnm commented on third_party/nvfuser/csrc/compute_at_map.cpp:
```diff
@@ -169,6 +169,15 @@ bool IterDomainGraph::exprsMap(
     }
   }
 
```
Am I understanding correct that, if I have `I1` and `I3` exact mapped, and
```
producer: I2 = resize(I1, l1, r1)
consumer: I4 = resize(I3, l2, r2)
```
If `l1->sameAs(l2) && r1->sameAs(r2)`, then

exact map, permissive map:
```
{I1, I3}
{I2, I4}
```

permissive resize map:
```
{I1, I2, I3, I4}
```

And if `!(l1->sameAs(l2) && r1->sameAs(r2))`, then

exact map, permissive map:
```
{I1, I3}
{I2}
{I4}
```

permissive resize map:
```
{I1, I2, I3, I4}
```

naoyam replied:
I'm not sure what you mean by `producer: I2` and `consumer: I4`, but assuming they are just produced from `I1` and `I3`, respectively, yes, that is correct.

zasdfgbnm replied:
I meant I have
```
Tconsumer = pad(Tproducer)
```
where
```
Tproducer:
root domain: I1
leaf domain: I2
transformations: I2 = resize(I1, l1, r1)

Tconsumer
root domain: I3
rfactor domain: I4
leaf domain: I4
transformations: I4 = resize(I3, l2, r2)
```

naoyam replied:
Assuming `transformations: I2 = resize(I1, l1, r1)` is not done as part of the rfactor transformation of `Tproducer`, it only happens when `Tproducer` is transformed as `Tconsumer` for producer indexing. In that case,while `I2` and `I4` should map, we don't update the ComputeAt map, so it won't affect how any of the exact/permissive/etc maps.

zasdfgbnm reviewed (commented):
posting more

Commit:
PR feedback

Commit:
Mark pad widths of PadOp as inputs

This is required to make the widths vals to segmented fusions

Also did some refactoring of PadOp accessors

Commit:
cleanup

Commit:
comments

Commit:
cleanpu

Commit:
Rename "disjoint view" to "disjoint rfactor"

Commit:
Set the producer and consumer maps of the ID graph using the
permissive-resize map

Commit:
Do not create index maps for non-indexable domains

Commit:
cleanup

Commit:
fix dependency analysis

Commit:
Naming fix

csarofeen commented on third_party/nvfuser/csrc/ir_internal_nodes.h:
```diff
@@ -2181,4 +2238,141 @@ class TORCH_CUDA_CU_API NamedScalar : public Val {
+};
+
+// Similar to at::indexing::Slice
+struct Slice {
```
Just a note for now: We should build out specialization in the cache when start/stop/step are trivial cases. Similar to how we have caching associated with view transformations.

csarofeen replied:
I think this is going to be important for piping through to the front end.

csarofeen commented on third_party/nvfuser/csrc/ir_nodes.cpp:
```diff
@@ -2127,6 +2129,52 @@ std::pair<IterDomain*, IterDomain*> IterDomain::swizzle(
+
+  TORCH_CHECK(
+      in->start()->isZeroInt(),
+      "Non-zero start not considered: ",
```
nit: considered -> supported

csarofeen commented on third_party/nvfuser/csrc/ir_nodes.cpp:
```diff
@@ -2996,4 +3075,211 @@ c10::optional<ParallelType> NamedScalar::getParallelIndex() const {
+int PadOp::getNumPaddedAxes() const {
+  int num_pad_inputs =
+      std::distance(getPadWidthInputBegin(), getPadWidthInputEnd());
+  TORCH_INTERNAL_ASSERT(num_pad_inputs % 2 == 0);
```
Shouldn't this check be done in the constructor on pad_widths?

csarofeen commented on third_party/nvfuser/csrc/ir_internal_nodes.h:
```diff
@@ -2181,4 +2238,141 @@ class TORCH_CUDA_CU_API NamedScalar : public Val {
+
+  //! Pad a tensor as specified by a vector of integer scalars. For
+  //! the actual semantics, see the torch.pad documentation
+  PadOp(
```
What's the syntax here? Having trouble decifering pad_widths usage. Seems you're assuming there's a number of entries not necessarily matching the input num dimensions * 2?

csarofeen replied:
I saw your comment in alias.h, can you add a size assert in the constructor here?

csarofeen commented on third_party/nvfuser/csrc/ir_internal_nodes.h:
```diff
@@ -2181,4 +2238,141 @@ class TORCH_CUDA_CU_API NamedScalar : public Val {
+      IrBuilderPasskey passkey,
+      TensorView* out,
+      TensorView* inp,
+      const std::vector<Slice>& ranges);
```
Similar question to pad op, what's the syntax for ranges? Seems it's not necessarily that ranges.size() == inp->nDims()?

naoyam replied:
I forgot validating the size of the `ranges` vector. It should be the same as `inp->nDims()`.

csarofeen commented on third_party/nvfuser/csrc/ir_utils.h:
```diff
@@ -364,5 +364,8 @@ TORCH_CUDA_CU_API bool isTorchGatherLookupTv(const Val* tv);
 TORCH_CUDA_CU_API std::string varName(const Val* val);
 
+// Check if a tensor is resized as part of  its root to rfactor transformations
+bool hasResizedRfactor(const TensorView* tv);
```
Is it true/enforced somewhere that resize only exists between root and rfactor?

naoyam replied:
Resize can exist after rfactor. Suppose:

```
tv1 = pad(tv0, {...});
```

And when indexing this expression, `tv0` is transformed as `tv1`, which includes replaying the resize expr of `tv1`. So it can exist outside of the rfactor transformation, but it needs to be done carefully. Resizing a producer tensor is fine as the transformation is temporary and it doesn't affect how it is actually calculated. However, we shouldn't use `resize` as a scheduling primitive like `merge` and `split`, and there's no validation for that.

I think what we could do is to do some validation at the beginning of lowering. At that point, resize should only be used between root and rfactor.

naoyam replied:
Related comment: https://github.com/csarofeen/pytorch/pull/2480#discussion_r1122746605

naoyam replied:
Validation added: 3ed4797d13e34048ef5ffae15786cb8d9164dd6d

csarofeen replied:
Makes sense, but scares me in the way that it also feels a lot like broadcast resolution.

csarofeen commented on third_party/nvfuser/csrc/index_compute.h:
```diff
@@ -320,12 +321,20 @@ class Index {
+      TensorView* producer,
+      const TensorView* consumer,
+      const std::vector<kir::ForLoop*>& loops,
+      const std::unordered_set<kir::ForLoop*>& rotated_loops,
```
I didn't review the loop rotation PR, but is it possible to model it for indexing in iter domain exprs? Similar question for the override_index entry.

naoyam replied:
For override_index, I think that's possible. Something like scatter and gather as IterDomain expressions is what I'm thinking about.

For loop rotation, that's probably possible, but not sure how beneficial it would be. Looks like much of the complexity comes from actual loop generation and indexing, and I don't know if representing it as an IterDomain expr would be helpful.

csarofeen replied:
I think it would be good to try and keep indexing as self contained as possible. Having these arguments passed in like this make me worried about trying to refactor indexing. Can revisit later.

csarofeen commented on third_party/nvfuser/csrc/ir_nodes.cpp:
```diff
@@ -2996,4 +3075,211 @@ c10::optional<ParallelType> NamedScalar::getParallelIndex() const {
+    IrBuilderPasskey passkey,
+    TensorView* out,
+    TensorView* inp,
+    const std::vector<Val*>& pad_widths)
```
Can you add an assert here that pad_widths is <= input dims (without reduction dims) * 2 and pad_widths % 2 == 0? I think you might be doing this in alias.cpp, but should be here too.

csarofeen commented on third_party/nvfuser/csrc/ops/alias.cpp:
```diff
@@ -371,4 +372,285 @@ TensorView* transpose(TensorView* x) {
+// Padding widths are assumed to be non-negative. Currently there's no
+// validation.
+TensorView* pad(TensorView* inp, const std::vector<Val*>& pad_widths) {
+  const auto& inp_dom = inp->domain()->noReductions();
```
should we stick with inp->getMaybeRFactor instead of inp->domain? Shouldn't make a real difference, and maybe it's just convention of the file.

naoyam replied:
Yeah, it shouldn't matter as this function is supposed to be used as part of fusion definitions, so `inp->domain()` should be the same as `inp->getMaybeRFactorDomain()`. That said, I prefer to follow the convention. Changed.

csarofeen commented on third_party/nvfuser/csrc/ops/alias.cpp:
```diff
@@ -371,4 +372,285 @@ TensorView* transpose(TensorView* x) {
+  std::vector<IterDomain*> rfactor_ids(ndims);
+
+  // PadOp requires pad widths for all dimensions, even for non-padded
+  // ones.
```
This doesn't seem true to me based on PadOp impl. Are you sure?

naoyam replied:
For avoiding ambiguity, I decided to make the padding vector of `PadOp` always have the width vals for all dimensions, no matter if each dimension is actually padded. Added error checks to `PadOp` constructor and methods.

csarofeen commented on third_party/nvfuser/csrc/ops/alias.cpp:
```diff
@@ -371,4 +372,285 @@ TensorView* transpose(TensorView* x) {
+  }
+
+  // torch.pad has padding widths of inner dimensions before outer
+  // dimensions
```
🤢 I dislike how the Torch API has these reversed. Thanks for not maintaining this convention in PadOp.

csarofeen commented on third_party/nvfuser/csrc/ops/alias.cpp:
```diff
@@ -371,4 +372,285 @@ TensorView* transpose(TensorView* x) {
+    auto inp_root_id = inp_dom[idx];
+    IterDomain* out_root_id = nullptr;
+    IterDomain* out_rf_id = nullptr;
+    if (idx < num_non_padded_dims) {
```
Can you be a little more precise with an entry filled in that's zero padded? Just because entries were passed in, doesn't mean they're actually padded.

naoyam replied:
Updated not to use `resize` when both left and right pad widths are zero.

csarofeen commented on third_party/nvfuser/csrc/ops/alias.cpp:
```diff
@@ -371,4 +372,285 @@ TensorView* transpose(TensorView* x) {
+// cat is implemented as PadOp and CatOp. Padding is done first to
+// account for the size difference between each of the inputs and the
+// output. All of the inputs to CatOp have the same shape as the
+// output shape.
```
So all inputs to cat op are padded first? then it's simply a predicate that selects between the different tensors?

csarofeen replied:
I think this understanding lines up with codegen.cpp

csarofeen replied:
Oh, it's not actually a "pad" but a resize, right?

csarofeen replied:
Inserting a pad op here seems strange to me. The mismatch in extents should be easily detected by the resize operation. What does inserting the pad op do here? I guess you're just inserting an operation so there's an opportunity for the inputs to have an rfactor domain which contains that resize.

naoyam replied:
Padding is done to resize each input so that all inputs are properly lined up with the output. Each input needs to be resized individually since the padding widths are specific to each input, e.g., the first input tensor has zero left width and right width of `(total_extent - extent_of_first_tensor_extent)`. 

This way, each input as a producer is trivially indexed. If there's no `PadOp` between the `cat` output and inputs, I'm not sure what the output IterDomain should look like. We know its extent, but to express it as a resize output, we also need to determine the left and right expansion factors, which isn't unique depending on which input tensor is considered as the input of the resize operation.

naoyam replied:
Having `PadOp` makes `CatOp` work pretty seamlessly, but I should have mentioned that there's one downside that we end up materializing the padding, so the generated code isn't ideal. For example, here's an excerpt from `FusionResizeCat4`:

```
  float T3[1];
  if ((i124 < ((T0.size[0] * T1.size[1]) + i360))) {
    T3[0]
       = ((((((((nvfuser_index_t)blockIdx.x) * 128) + ((nvfuser_index_t)threadIdx.x)) % (T1.size[1] + i7)) - i7) >= 0) && (((((((nvfuser_index_t)blockIdx.x) * 128) + ((nvfuser_index_t)threadIdx.x)) % (T1.size[1] + i7)) - i7) < T1.size[1])) ? T1[(((i124 / i125) * T1.stride[0]) + (((i124 % i125) - T0.size[1]) * T1.stride[1]))] : 0.000000000e+00f;
  }
  float T2[1];
  if (b407) {
    T2[0]
       = ((((((((nvfuser_index_t)blockIdx.x) * 128) + ((nvfuser_index_t)threadIdx.x)) % (T0.size[1] + i6)) - 0) >= 0) && (((((((nvfuser_index_t)blockIdx.x) * 128) + ((nvfuser_index_t)threadIdx.x)) % (T0.size[1] + i6)) - 0) < T0.size[1])) ? T0[(((i124 / i259) * T0.stride[0]) + (((i124 % i259) - 0) * T0.stride[1]))] : 0.000000000e+00f;
  }
  if (b407) {
    if (((((((nvfuser_index_t)blockIdx.x) * 128) + ((nvfuser_index_t)threadIdx.x)) % (T0.size[1] + i6)) < (0 + T0.size[1]))) {
      T4[i124] = T2[0];
    } else {
      T4[i124] = T3[0];
    }
  }
```

Kernel math:

```
%kernel_math {
i5 = i2 + i4;
i6 = i5 - i2;
T2_l[ iS22{( ceilDiv(( i0 * ( i2 + ( ( i2 + i4 ) - i2 ) ) ), 128) )}, iS23{128} ] ca_pos( 2 )
   = pad( T0_g[ iS25{( ceilDiv(( i0 * i2 ), 128) )}, iS26{128} ], {0, 0, 0, i6} )
i7 = 0 + i2;
T3_l[ iS16{( ceilDiv(( i3 * ( i4 + ( 0 + i2 ) ) ), 128) )}, iS17{128} ] ca_pos( 2 )
   = pad( T1_g[ iS19{( ceilDiv(( i3 * i4 ), 128) )}, iS20{128} ], {0, 0, i7, 0} )
T4_g[ iblockIdx.x13{( ceilDiv(( i0 * ( i2 + ( ( i2 + i4 ) - i2 ) ) ), 128) )}, ithreadIdx.x14{128} ] ca_pos( 2 ) produce_pos( 2 )
   = cat( T2_l[ iS22{( ceilDiv(( i0 * ( i2 + ( ( i2 + i4 ) - i2 ) ) ), 128) )}, iS23{128} ] ca_pos( 2 ), T3_l[ iS16{( ceilDiv(( i3 * ( i4 + ( 0 + i2 ) ) ), 128) )}, iS17{128} ] ca_pos( 2 ), 1 )
}
```

Ideally, this should just be:

```
  if (b407) {
    if (((((((nvfuser_index_t)blockIdx.x) * 128) + ((nvfuser_index_t)threadIdx.x)) % (T0.size[1] + i6)) < (0 + T0.size[1]))) {
      T4[i124] = T0[(((i124 / i259) * T0.stride[0]) + (((i124 % i259) - 0) * T0.stride[1]))];
    } else {
      T4[i124] =  T1[(((i124 / i125) * T1.stride[0]) + (((i124 % i125) - T0.size[1]) * T1.stride[1]))];
    }
  }
```

I think this translation could be done after lowering at the Kernel IR level. Will revisit later.

csarofeen replied:
Yeah, that's kind of what I was worried about/thinking. Makes sense as a follow up, or maybe we can chat today in our meeting if there's another way to model this (not really sure there is).

csarofeen replied:
Still want to follow up on this, please remind me in our next codegen meeting.

csarofeen reviewed (commented):
I'm not finished reviewing yet, pushing some comments...

I think these ops could have interesting consequences for caching. It seems worthwhile to talk out how these ops might be cached in the fusion definition for specialization (especially to capture trivial instances).

We might want to generalize how we cache based on transformations between root and rfactor in view transformations to generally consist of all transformations between all tensor views between root and rfactor domains.

It would be nice if we added some more information to some of the asserts in the PR, especially when an operator fails due to user input it would be good to mention which operation and more details about which input causes the issue.

csarofeen commented on third_party/nvfuser/csrc/ops/alias.cpp:
```diff
@@ -371,4 +372,285 @@ TensorView* transpose(TensorView* x) {
+
+  // Special handling for the case where there's only one input
+  if (inputs.size() == 1) {
+    return set(inputs.at(0));
```
Would be good to have a similar shortcut in padding and slice, if we get a detectably trivial case.

naoyam replied:
Updated `pad`. There's already a shortcut in `slice`.

Commit:
Use pad instead of manually creating PadOp

Commit:
Remove unused func

Commit:
remove unnecessary func

Commit:
PR feedback

zasdfgbnm reviewed (commented):
I think all my concerns have been resolved. This PR LGTM now. I am not stamping to avoid accidental merge because @csarofeen hasn't done review yet. Please let me know if you want a stamp.

Commit:
Clean up pad_width vector

Commit:
Convert pad to set when valid

Commit:
More descriptive error messages

naoyam commented:
> It would be nice if we added some more information to some of the asserts in the PR, especially when an operator fails due to user input it would be good to mention which operation and more details about which input causes the issue.

I revisited them and added more descriptive error messages.

Commit:
Merge branch 'devel' into iter_domain_resize

Commit:
Missing override implementation

Commit:
test fix

Commit:
Add validation of resize usage

Commit:
rename test file

naoyam commented:
@zasdfgbnm I added this commit that adds an override for `Resize`: f187d6d. This is only used when resize is used with a tensor that has siblings, and since there's no such expr, it's just for consistency. I also noticed there's no override for `Swizzle2D`, which is fine as there shouldn't be any multi-output expr whose outputs should be swizzled, but added an assert just in case.

csarofeen commented on third_party/nvfuser/csrc/lower_validation.cpp:
```diff
@@ -1337,4 +1337,25 @@ void validateLookupTV(Fusion* fusion) {
+            [](Expr* expr) { return expr->isA<Resize>(); }),
+        "Invalid use of resize detected with ",
+        tv->toString(),
+        ". Resize may only be used as part of rfactor transformations.");
```
I thought you said there could be resizes in produces replayed like consumers that are not within rfactor.

naoyam replied:
Yes, that should only happen during lowering when a producer is replayed as consumer for producer indexing. So, at this point, there should never be resize after rfactor.

csarofeen commented on third_party/nvfuser/csrc/scheduler/utils.h:
```diff
@@ -557,5 +561,19 @@ inline void rotateLoop(
+//! move its producer to a shared memory that is sufficient to satisfy
+//! the dependency. A proper RAW sync will be automatically inserted
+//! when the fusion is lowered.
+TORCH_CUDA_CU_API void promoteProducerMemoryTypesOfResizedTensors(
```
Should this be part of the heuristics in case shared memory isn't large enough to do this?

naoyam replied:
Agreed, but not sure what the right heuristics should look like. If there's not large enough shared memory buffer is available, an easiest workaround would be using global memory, but that could be a big perf overhead. Should we instead consider different parallelization that won't require a RAW sync? To develop an effective heuristic, we would need to see real instances. I'll keep it as is for now. If the buffer space becomes a problem, we should see a hard error before a kernel is launched.

csarofeen commented on third_party/nvfuser/csrc/index_compute.cpp:
```diff
@@ -582,8 +582,38 @@ void IndexCompute::handle(Swizzle2D* swizzle_2d) {
+    // resize is not indexed as the indexing is done using the rfactor root
+    // domain. This could be an issue when a resize is shows up outside of
+    // rfactor transfomations, but currently that only can happen when a
+    // producer tensor is transformed to look like a consumer. Since inlining is
```
Okay, this was what I was missing on having resize outside root->rfactor.

csarofeen commented on third_party/nvfuser/test/test_resize.cpp:
```diff
@@ -0,0 +1,1594 @@
+}
+
+// Disabled due to the same reason as Pad8
+#if 0
```
Shouldn't this case just get segmented so the multiple resizes don't appear in the same kernel?

csarofeen replied:
I don't think we can enable support if we don't segment properly on all cases.

naoyam replied:
An interesting part of this fusion is, although `tv1` is padded twice differently, the two padding outputs, `tv2` and `tv3`, are summed together, producing `tv4`, so the rfactor IDs of `tv2`, `tv3` and the root IDs of `tv4` are all exactly mapped. If not, multiple different resizes ops should result in segmented fusions, much like different views are handled.

This test actually works successfully. It's just scheduled as a single pointwise fusion. But I'm not sure how a fusion like this should be scheduled.

csarofeen replied:
So the point being root and rfactor match exactly because producers and consumers are common, but the root->rfactor paths are different. That's interesting.

naoyam replied:
That's right.

csarofeen commented on third_party/nvfuser/test/test_resize.cpp:
```diff
@@ -0,0 +1,1594 @@
+}
+
+// Cat many tensors
+TEST_F(NVFuserTest, FusionResizeCat7_CUDA) {
```
Do you check that an error gets thrown if runtime values don't support the requested concat?

naoyam replied:
The type and number of dimensions must match:
 https://github.com/csarofeen/pytorch/pull/2480/files/82845fa11d79a153f62fd11fa4daf473476cd728#diff-bee9b138517ab66ac5e215f0603aacbacc8984fda748a6d636e36595ec0b12deR467-R494.

But that's it. For example, the extents of the non-concatenated dimensions must match, but that's not checked, like normal arithmetic ops where we assume two operands have the same shape.

csarofeen replied:
I think it makes sense to add a runtime check for this.

naoyam replied:
I'm not sure what to check. We don't check, for example, in `add(tv0, tv1)`, `tv0` and `tv1` have the same shape, but rather we assume that's the case and use that to build the IterDomain graph. Following the same approach, non-concatenated domains should be assumed to be the same.

csarofeen replied:
Heh, that's a good point, and that's why this would error out during expression evaluator as our evaluated symbols for parallelization wouldn't match. Okay, will put it in the large bucket of things we should do. Would be nice to have a validation check that everything matches as expected across the iter domain relationships.

naoyam replied:
Something like a pedantic mode, where everything is validated no matter how high its overhead would be.

csarofeen commented on third_party/nvfuser/test/test_resize.cpp:
```diff
@@ -0,0 +1,1594 @@
+  auto tv5 = sum(tv3, {1});
+  auto tv6 = sum(tv4, {1});
+  auto tv7 = cat({tv5, tv6}, 0);
+  fusion.addOutput(tv7);
```
Are you thinking register+smem persistent kernels here??

naoyam replied:
Yes. That's still a big gap I really wanted to fill.

csarofeen reviewed (approved):
Finished reviewing everything. I don't see any blockers. Nice work.

Commit:
fix

Commit:
Merge branch 'devel' into iter_domain_resize

Commit:
merge fix

Commit:
test fix

jjsjann123 commented:
I'm getting segfault from this PR. https://github.com/NVIDIA/Fuser/pull/3#issuecomment-1467749444

naoyam commented:
> I'm getting segfault from this PR. [NVIDIA/Fuser#3 (comment)](https://github.com/NVIDIA/Fuser/pull/3#issuecomment-1467749444)

Hmm, strange. Looking into it.
